Firestore/core/src/firebase/firestore/core/firestore_client.mm

```diff
@@ -94,18 +94,20 @@
       new FirestoreClient(database_info, std::move(credentials_provider),
                           std::move(user_executor), std::move(worker_queue)));
 
-  auto user_promise = std::make_shared<std::promise<User>>();
   bool credentials_initialized = false;
 
   std::weak_ptr<FirestoreClient> weak_client(shared_client);
-  auto credential_change_listener = [credentials_initialized, user_promise,
-                                     weak_client](User user) mutable {
+  auto credential_change_listener = [credentials_initialized, weak_client,
+                                     settings](User user) mutable {
     auto shared_client = weak_client.lock();
     if (!shared_client) return;
 
     if (!credentials_initialized) {
       credentials_initialized = true;
-      user_promise->set_value(user);
+
+      shared_client->worker_queue()->Enqueue([shared_client, user, settings] {
```

Comment:
This doesn't work.

The whole point of enqueueing outside the credential change listener is that we want to prevent any API call from getting on the async queue before we have the current identity. 

The `user_promise->get_future().get()` blocks in the body of that callback essentially preventing the async queue from doing anything until we've gotten the callback.

With this change, any API call can now proceed before Auth has called us back and will hit the internals before we're initialized.

Reply:
As discussed offline, this is actually safe since `SetCredentialChangeListener` invokes the callback synchronously on the calling thread when it is first registered. Thus, the client initialization continues to be the first item enqueue on the worker queue.

I added a comment and an assert that makes this more obvious. As part of this, I also had to change credentials_initialized tp be captured by reference.